csds-client/client/v3/client.go

```diff
@@ -0,0 +1,390 @@
+
+	req := &csdspb_v3.ClientStatusRequest{NodeMatchers: c.nodeMatcher}
+	if err := streamClientStatus.Send(req); err != nil {
+		return err
+	}
+
+	resp, err := streamClientStatus.Recv()
+	if err != nil && err != io.EOF {
+		return err
+	}
+	// post process response
+	if err := printOutResponse(resp, c.opts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// parseConfigStatus parses each xds config status to string
+func parseConfigStatus(xdsConfig []*csdspb_v3.PerXdsConfig) []string {
+	var configStatus []string
+	for _, perXdsConfig := range xdsConfig {
+		status := perXdsConfig.GetStatus().String()
+		var xds string
+		if perXdsConfig.GetClusterConfig() != nil {
+			xds = "CDS"
+		} else if perXdsConfig.GetListenerConfig() != nil {
+			xds = "LDS"
+		} else if perXdsConfig.GetRouteConfig() != nil {
+			xds = "RDS"
+		} else if perXdsConfig.GetScopedRouteConfig() != nil {
+			xds = "SRDS"
+		} else if perXdsConfig.GetEndpointConfig() != nil {
+			xds = "EDS"
+		}
+		if status != "" && xds != "" {
+			configStatus = append(configStatus, xds+"   "+status)
+		}
+	}
+	return configStatus
+}
+
+// printOutResponse processes response and print
+func printOutResponse(response *csdspb_v3.ClientStatusResponse, opts client.ClientOptions) error {
+	if response.GetConfig() == nil || len(response.GetConfig()) == 0 {
+		fmt.Printf("No xDS clients connected.\n")
+		return nil
+	} else {
+		fmt.Printf("%-50s %-30s %-30s \n", "Client ID", "xDS stream type", "Config Status")
+	}
+
+	var hasXdsConfig bool
+
+	for _, config := range response.GetConfig() {
+		var id string
+		var xdsType string
+		if config.GetNode() != nil {
+			id = config.GetNode().GetId()
+			metadata := config.GetNode().GetMetadata().AsMap()
+
+			// control plane is expected to use "XDS_STREAM_TYPE" to communicate
+			// the stream type of the connected client in the response.
+			if metadata["XDS_STREAM_TYPE"] != nil {
+				xdsType = metadata["XDS_STREAM_TYPE"].(string)
+			}
+		}
+
+		if config.GetXdsConfig() == nil {
+			if config.GetNode() != nil {
+				fmt.Printf("%-50s %-30s %-30s \n", id, xdsType, "N/A")
+			}
+		} else {
+			hasXdsConfig = true
+
+			// parse config status
+			configStatus := parseConfigStatus(config.GetXdsConfig())
+			fmt.Printf("%-50s %-30s ", id, xdsType)
+
+			for i := 0; i < len(configStatus); i++ {
+				if i == 0 {
+					fmt.Printf("%-30s \n", configStatus[i])
+				} else {
+					fmt.Printf("%-50s %-30s %-30s \n", "", "", configStatus[i])
+				}
+			}
+			if len(configStatus) == 0 {
+				fmt.Printf("\n")
+			}
+		}
+	}
+
+	if hasXdsConfig {
+		if err := client.PrintDetailedConfig(response, opts); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// parseYaml is a helper method for parsing csds request yaml to nodematchers
+func parseYaml(path string, yamlStr string, nms *[]*envoy_type_matcher_v3.NodeMatcher) error {
+	if path != "" {
```

Comment:
Feels like there is some code that is generic enough to move it to utils and not repeat for each version, what do you think?

Reply:
I've extracted `ParseYamlFileToMap()` and `ParseYamlStrToMap` to clientUtil